src/compiler/checker.ts

```diff
@@ -22914,7 +22914,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function isTupleLikeType(type: Type): boolean {
-        return isTupleType(type) || !!getPropertyOfType(type, "0" as __String);
+        let lengthType;
+        return isTupleType(type) ||
+            !!getPropertyOfType(type, "0" as __String) ||

```

Comment:
I wonder how much it'd take to eliminate this check; shouldn't all tuples have a length property?

Reply:
Do you mean the "0" prop check?

Reply:
Yes, though, judging by the baseline changes there are some other things that would need to be fixed, so maybe it's unwise to try.

Reply:
Yeah, you have things like this in the test suite:
https://github.com/microsoft/TypeScript/blob/1ed0a5ac4b49971e4698d5388f7760cbaa0615ac/tests/cases/compiler/inferringAnyFunctionType1.ts

I think I might have seen `0` prop being added to things in the wild to hint at tuple inference and stuff. Arguably, they were abusing this particular heuristic a little bit - but it's also not like it was hidden (I've seen it mentioned on a couple of occasions, including in the linked ticket [here](https://github.com/microsoft/TypeScript/issues/29311#issuecomment-453337774))

Reply:
@jakebailey any further thoughts about this PR? I've recently learned that `'0'` thing was often~ used to imitate tuples when tuples were supported by TS

Reply:
Getting to this later than I intended; I don't think I have any remaining concerns here, though the code flow of `lengthType` is a little funky compared to a plain if statement. Not sure it matters. I bet there's some silly v8 thing where the expression is faster for some silly reason.

Reply:
Hm, i wouldnt thought about this being faster (maybe it is, i dont know). It’s not exactly how i would write it - it’s just using a similar code style that is used in other places in the codebase.